### PR TITLE
[http2_adapter] HTTP/2 proxy

### PR DESCRIPTION
**Pull request by Overman775**
### New Pull Request Checklist

- [X] I have read the [Documentation](https://pub.dartlang.org/packages/dio)
- [X] I have searched for a similar pull request in the [project](https://github.com/flutterchina/dio/pulls) and found none
- [X] I have updated this branch with the latest `develop` to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I am adding
- [X] I have updated the documentation (if necessary)
- [X] I have run the tests and they pass

This merge request fixes / refers to the following issues: #1259 #905

### Pull Request Description

Added proxy tunnel to http2 adaptor

### CODE REVIEW

File: plugins/http2_adapter/lib/src/client_setting.dart
Comment (ueman) [inline plugins/http2_adapter/lib/src/client_setting.dart:17]: Please add some documentation here, because I'm kinda lost what to do with it.

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (ueman) [inline plugins/http2_adapter/lib/src/connection_manager_imp.dart:176]: Is the error case here intentionally not handled? 
I believe exceptions should be caught and rethrown as `DioError` like in the rest of the code base.

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (ueman) [inline plugins/http2_adapter/lib/src/connection_manager_imp.dart:137]: I'm not really familiar with how proxys work. Could you maybe write some comments on the general flow and why it's done like that?

**Comment by ueman** [comment]
@Overman775 You should really write tests covering these changes. Otherwise, somebody could break this functionality without noticing. 

Also please don't tick the checkbox which says that you've written tests, if you don't have written tests.

I'm also not a maintainer of this library, but I believe the maintainer would agree in most cases.

Overall it looks already promising and it would be a great addition to the library 👍

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (kuhnroyal) [inline plugins/http2_adapter/lib/src/connection_manager_imp.dart:76]: Since you are passing the `clientConfig` anyways, both cases can actually be collapsed into one method call and do the `if (clientConfig.proxy != null)` check in there.

**Review by kuhnroyal — CHANGES_REQUESTED** [review]
I am unfamiliar with how proxying does work with HTTP2 but I can probably test it once it is in a complete state.
Looks like it could work, nice addition!

File: plugins/http2_adapter/lib/src/client_setting.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/client_setting.dart:20]: @ueman added

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart:185]: Yes, all errors are caught above, but I added 'onError', is that better?

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart:137]: You can refer  https://www.ietf.org/rfc/rfc2817.txt point 5.2. We just create a socket(proxy tunnel), send headers and that's it, it's simple.

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart]: fixed and aded proxy authorization

**Comment by Overman775** [comment]
> @Overman775 You should really write tests covering these changes. Otherwise, somebody could break this functionality without noticing.
> 
> Also please don't tick the checkbox which says that you've written tests, if you don't have written tests.
> 
> I'm also not a maintainer of this library, but I believe the maintainer would agree in most cases.
> 
> Overall it looks already promising and it would be a great addition to the library 👍


There were no tests initially, it is difficult to write tests for Socket. This code does not contain breakin changes

**Comment by Overman775** [comment]
@ueman @kuhnroyal pleace check PR with test @yfdyh000

**Comment by chejdj** [comment]
Hello, could you see this pr? @wendux

**Review by AlexV525 — COMMENTED** [review]
Hi @Overman775. I left some comments in the PR.

Also, there are a few things I want to mention here:
- Are you willing to contribute by letting us pick the PR to another community-maintained `dio`? (We'll announce the new repo later once we merge PRs as much as possible.)
- It might be better to add some tests.

**Comment by AlexV525** [comment]
@Overman775 Are you still able to work on this?

**Comment by Overman775** [comment]
> @Overman775 Are you still able to work on this?

Thought it would never happen ) I will update the PR soon

**Comment by AlexV525** [comment]
> Thought it would never happen ) I will update the PR soon

For now you can consider to purpose the PR to [diox](https://github.com/cfug/diox) since dio is not going to maintain anymore.

**Comment by AlexV525** [comment]
Or you can hold on for a couple of days, we have some incoming changes that need to announce. :)

**Comment by talisk** [comment]
> > @Overman775 Are you still able to work on this?
> 
> Thought it would never happen ) I will update the PR soon



> Or you can hold on for a couple of days, we have some incoming changes that need to announce. :)

Now dio will continue to be maintained by the CFUG organization, diox has been migrated to dio, you can continue your work and join in dio.

File: plugins/http2_adapter/README.md
Comment (Overman775) [reply plugins/http2_adapter/README.md]: rewrited same as https://github.com/cfug/dio/blob/main/example/lib/http2_adapter.dart and added proxy doc

File: plugins/http2_adapter/example/example.dart
Comment (Overman775) [reply plugins/http2_adapter/example/example.dart]: rewrited same as https://github.com/cfug/dio/blob/main/example/lib/http2_adapter.dart

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart]: refactored because lint [rule](https://dart-lang.github.io/linter/lints/void_checks.html) give warning

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart]: not actual, changes reverted

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart:159]: tested on Windows and Mac, works fine)

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart]: refactored

Comment (Overman775) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart]: but we can't return as above because it is necessary to close `proxySubscription` after creating `SecureSocket`

**Comment by Overman775** [comment]
test added )

**Comment by vento007** [comment]
will this work on web as well?

**Comment by Overman775** [comment]
> will this work on web as well?

this plugin don't support web

**Comment by simonpojok** [comment]
Hey, guys I wanted to get started supporting and contributing to this package, I need someone to onboard me. You can DM me directly on simonpojok@gmail.com

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (AlexV525) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart:150]: > No, it doesn't depend on the platform, proxy connection uses ISO-8859-1 encoding

Can you link the standard here for better understanding?

**Comment by AlexV525** [comment]
> Test added. the test requires a proxy server, do you have any ideas how to check?

Do we have a minimum requirement for setting up an HTTP/2 proxy?

**Comment by Overman775** [comment]
> Do we have a minimum requirement for setting up an HTTP/2 proxy?

simple http\https proxy with normal latency

**Comment by AlexV525** [comment]
> simple http\https proxy with normal latency theoretically, we can run a local proxy server on github actions

Sounds good. You can follow how the server is being setup in the existing tests to write startup the proxy in the proxy tests.

**Comment by Overman775** [comment]
> > simple http\https proxy with normal latency theoretically, we can run a local proxy server on github actions
> 
> Sounds good. You can follow how the server is being setup in the existing tests to write startup the proxy in the proxy tests.

all done, added proxy server on test flow. Please approve new test actions https://github.com/cfug/dio/actions/runs/4317864902

File: plugins/http2_adapter/lib/src/connection_manager_imp.dart
Comment (AlexV525) [reply plugins/http2_adapter/lib/src/connection_manager_imp.dart:150]: This can be improved by adding the ISO-8859-1 link as a comment.

**Comment by Overman775** [comment]
@AlexV525 When will the request be approved? Is there anything else that needs to be finalized?

**Comment by AlexV525** [comment]
Ah so you indeed didn't notice the previous comment as I imagine. See https://github.com/cfug/dio/pull/1386#discussion_r1128730582

**Comment by Overman775** [comment]
> Ah so you indeed didn't notice the previous comment as I imagine. See [#1386 (comment)](https://github.com/cfug/dio/pull/1386#discussion_r1128730582)

@AlexV525 done )

**Review by AlexV525 — APPROVED** [review]
Thanks for the great feature. cc @kuhnroyal since you've requested changes.

**Review by kuhnroyal — APPROVED** [review]
Hmm, yea that was over a year ago. But looks good!

**Pull request merged**
